byron/ledger/executable-spec/src/Ledger/Delegation.hs

```diff
@@ -575,3 +589,124 @@ initialEnvFromGenesisKeys ngk chainLength =
+delegatorDelegate = delegator &&& delegate
+
+ratioInt :: Int -> Int -> Double
+ratioInt x y
+  = fromIntegral x / fromIntegral y
+
+-- | Filters the list with the predicate and returns
+-- the ratio of filtered to original list lengths.
+lenRatio :: ([a] -> [b]) -> [a] -> Double
+lenRatio p xs
+  = ratioInt (length (p xs))
+             (length xs)
+
+maxInt :: [Int] -> Int
+maxInt [] = 0
+maxInt xs = List.maximum xs
+
+-- | True if the tuple count (snd item) is >= 2
+multiple :: (a,Int) -> Bool
+multiple = (2 <=) . snd
+
+-- | Count the number of delegates in the given certificates
+delegateCounts :: [DCert] -> [(VKey, Int)]
+delegateCounts certs
+  = fmap delegatorDelegate certs
+    -- Remove duplicated elements, since we're not
+    -- interested in repeated delegations
+    & List.nub
+    -- Select the (unique) delegates
+    & fmap snd
+    -- If we have more than one occurrence of a delegate,
+    -- then we know that there were multiple delegations
+    -- to that delegate
+    & count
+
+-- | Count the number of delegators in the given certificates
+delegatorCounts :: [DCert] -> [(VKeyGenesis, Int)]
+delegatorCounts dcerts
+  = fmap delegatorDelegate dcerts
+    -- Remove duplicated elements, since we're not
+    -- interested in repeated delegations
+    & List.nub
+    -- Select the (unique) delegators
+    & fmap fst
+    -- If we have more than one occurrence of a delegator,
+    -- then we know that the delegator changed their delegation
+    & count
+
+-- | Ratio of certificate groups that are empty
+emptyDelegationPayloadRatio :: [[DCert]] -> Double
+emptyDelegationPayloadRatio
+  = lenRatio (filter null)
+
+-- | Ratio of certificates that delegate to _this_ epoch, where
+--   each certificate is represented by (current epoch,cert epoch)
+thisEpochDelegationsRatio :: [(Epoch, Epoch)] -> Double
+thisEpochDelegationsRatio
+  = lenRatio (filter thisEpoch)
+  where
+    thisEpoch = uncurry (==)
+
+-- | Ratio of certificates that delegate to the _next_ epoch, where
+--   each certificate is represented by (current epoch,cert epoch)
+nextEpochDelegationsRatio :: [(Epoch, Epoch)] -> Double
+nextEpochDelegationsRatio
+  = lenRatio (filter nextEpoch)
+  where
+    nextEpoch (e0, e1) = e0 + 1 == e1
+
+-- | Ratio of certificates that "delegate to self", that is,
+-- where the delegator and delegate are the same
+selfDelegationsRatio :: [DCert] -> Double
+selfDelegationsRatio
+  = lenRatio (filter selfDeleg . fmap delegatorDelegate)
+  where
+    selfDeleg (vks, vk) = owner vks == owner vk
+
+-- | Ratio of delegates that have multiple delegators
+-- that are delegating to them
+multipleDelegationsRatio :: [DCert] -> Double
+multipleDelegationsRatio dcerts
+  = lenRatio (filter multiple) (delegateCounts dcerts)
+
+-- | The maximum number of delegators to any particular delegate
+maxDelegationsTo :: [DCert] -> Int
+maxDelegationsTo dcerts
+  =  maxInt (snd <$> filter multiple (delegateCounts dcerts))
```

Comment:
What does this give in case everyone has exactly one delegator?

Reply:
Good catch. It seems this will return 0 in that case, which is not right :)